### PR TITLE
Release Candidate V0.4

### PR DESCRIPTION
**Pull request by darkdrgn2k**
To fix

- [x] ./install2: line 18: warning: command substitution: ignored null byte in input (install2)
- [x] E: Unable to locate package armbian-firmware-full - pipe to dev null
- [x] IPSF spelled wrong on dialog
- [x] sed: cannot rename /tmp/sedKGXeLi: Operation not permitted [Possible solution](http://helpdoco.com/Linux-Unix/sed-cannot-rename-permission-denied.htm)
- [x] `for wlan in $(find /sys/class/net -maxdepth 1 -name 'wlan*' | xargs -0 basename); do` hostap fix
- [x] /var/www/html/cgj-bin/lib files incorrectly copy into /var/www/html/cgj-bin
- [x] Make selection screen a little bigger
- [x] Yggdrasil detection when not installed or errored out on map
- [x] Yggdrasil detection for status script when not installed

# Tested Install
- [x] Raspberry Pi 3 B - Full Node
- [x] Raspberry Pi 1 B - Basic Node
- [x] Orange Pi Zero 256 - Full Node
- [x] Rock64 - Basic Node
- [x] Nanopi Neo2 - Basic Node (Ubuntu)

# v0.4 Changelog
## Documentation
- Created `MODULES.md` documentation file
- Moved module related content from `README.md`
- Added documentation for various modules - more to come!
- Force password change on first login after install on Raspberry Pi

## Build
- Travis CI runs [shellcheck](https://www.shellcheck.net/) on all .sh and install files
- Created functions to minimize duplicate coding in install making install2 more readable
- Created Profiles for quick node configuration during install.
- Tweaks to build SSB Faster

## Modules
- CJDNS version bump
- Added [Yggdrasil](https://yggdrasil-network.github.io/) routing engine
- Added Yggdrasil IPv6 client IPs to access point ([Yggdrasil subnetting](https://yggdrasil-network.github.io/configuration.html#advertising-a-prefix))
- Added nginx welcome page with information about the node
- CJDNS and Yggdrasil live graphs of peers on welcome page
- Status script now work with Yggdrasil
- Added IP forwarding to iptunnel
- Added IPv6 addressing to iptunnel
- Sorted and cleaned up firewall rules
- Pi stream support for Audio Only (SDR)
- Pi stream patches for stability
- IPFS Version bump 0.4.18
- Added version variable and version bump for Grafana
- Added cjdns compiled DEB for arm64
- SSB Version bump and rename
- Yggdrasil data in Node Exporter

## Hardware
- Meshpoint and Adhoc dropped to 20Mhz 
- Raspbian onboard WiFi now easier to use in adhoc mode
- Added notes on how to mesh onboard Raspberry Pi WiFi
- Added install support for MK802ii
- Raspberry Pi Zero - enabled access point

## Other
- Lots of shellcheck suggested cleanup
- Scuttlebot now called ssb-server
- ipfs-swarm.sh now works with Yggdrasil too
- Added up forwarding rules for Yggdrasil client IPs
- Moved cjdns-setup from systemd rules to post startup script
- More IPFS bootstraps for Hyperborea and Yggdrasil
- IPFS  - enabled gossipsub and filestore 

## Contrib
- Added a port 80 redirect so non-Internet accessible nodes can show their Welcome page
    - Basically a captive gateway
- Added script to build SDR drivers for pi stream
- Added Hyperborea and Yggdrasil mirror

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -103,20 +76,21 @@ List of tested hardware:
...
+| Orange Pi One             | [Armbian](https://dl.armbian.com/orangepione/)                   | 160k,  74k | 67 Mbps | 1       | 10/100 | |
+| Orange Pi Zero            | [Armbian](https://dl.armbian.com/orangepizero/)                  | 160k,  74k | 67 Mbps | 1 (+2*) | 10/100 | *USB Headers |
+| Orange Pi Zero Plus 2 H5  | [Armbian](https://dl.armbian.com/orangepizeroplus2-h5/)          | 190k, 130K | 80 Mbps | 0 (+2*) | None   | *USB Headers |
+| NanoPi Neo 2              | [Armbian](https://dl.armbian.com/nanopineo2/)                    | 160k, 95K  | 67 Mbps | 1 (+2*) | 10/100/1000   | *USB Headers, Gigabit Eth |
+| Rock64                    | [Armbian](https://dl.armbian.com/rock64/)                        | 255k, 168K | 94 Mbps | 3       | 10/100/1000   | 1 USB 3.0, Gigabit Eth |
+| Pine64                    | [Armbian](https://dl.armbian.com/pine/nightly/)                  | 227k, 151k | 78 Mbps | 2       | 10/100/1000   |  Gigabit Eth |
+| EspressoBin               | [Armbian](https://dl.armbian.com/espressobin/)                   | 186k, 128K | 73 Mbps | 2       | 10/100/1000   | 1 USB 3.0, 3x Gigabit Eth, Sata, mPCIE. Use stable and apt-get upgrade after boot |
```
Comment (benhylau) [inline README.md]:
```suggestion
| ESPRESSObin               | [Armbian](https://dl.armbian.com/espressobin/)                   | 186k, 128K | 73 Mbps | 2       | 10/100/1000   | 1 USB 3.0, 3x Gigabit Eth, SATA, mPCIe. Use stable and apt-get upgrade after boot |
```

File: README.md [hunk 2]
```diff
@@ -103,20 +76,21 @@ List of tested hardware:
...
+| MK802ii       | Debian | 30k, 40k | 25Mbps | | | Android box. Single Core. Onboard wifi does Meshpoint |
```
Comment (benhylau) [inline README.md]:
```suggestion
| MK802ii       | Debian | 30k, 40k | 25Mbps | | | Android box. Single core. Onboard WiFi supports Mesh Point |
```

File: contrib/batman-adv/batman-adv-mesh.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+# Install batman adv
+sudo apt-get install -y batctl
+
+# configure batman-adv
```
Comment (benhylau) [inline contrib/batman-adv/batman-adv-mesh.sh]:
```suggestion
# Configure batman-adv
```

File: docs/FAQ.md [hunk 4]
```diff
@@ -105,6 +105,12 @@ saveenv
 ```
 ## Wireless
 
+**Q:** Why do my Meshpoint/AdHoc nodes Release v0.3 or lower no longer mesh with V0.4 or higher?
```
Comment (benhylau) [inline docs/FAQ.md]:
```suggestion
**Q:** Why do my MeshPoint/AdHoc nodes on v0.3 or lower no longer mesh with v0.4 or higher?
```

File: docs/FAQ.md [hunk 5]
```diff
@@ -105,6 +105,12 @@ saveenv
...
+
+**A:** We dropped the band width of MeshPoint and AdHoc to 20MHz from 40MHz. This should provider a bit better responsivness in urban environments.  Unfortunatlry the 20MHz and 40MHz bands do not work together.
```
Comment (benhylau) [inline docs/FAQ.md]:
```suggestion
**A:** We dropped the band width of MeshPoint and AdHoc to 20MHz from 40MHz in v0.4. This should provide a bit better responsiveness in urban environments.  Unfortunately the 20MHz and 40MHz bands do not work together.
```

File: docs/FAQ.md [hunk 6]
```diff
@@ -119,6 +125,9 @@ On board wireless we have seen so far
 To install
 * Install Ad-Hoc mesh module
 * Do NOT install Access Point
-* After install remove `HT40+` from `/usr/bin/mesh-adhoc`
 
 If you have success using ad-hoc with on board cards please let us know your experience.
```
Comment (benhylau) [inline docs/FAQ.md]: AdHoc vs. ad-hoc used across doc, consistent preferred.

File: scripts/hostapd/uninstall [hunk 7]
```diff
@@ -25,3 +25,4 @@ sudo rm -f /etc/hostapd/ca.pem
 sudo rm -f /etc/hostapd/server.pem
 sudo rm -f /etc/hostapd/nat.sh
 sudo rm -f /etc/systemd/system/hostapd.service
+sudo rm -rf /etc/udev/rules.d/70-persistent-net.rules || true
```
Comment (benhylau) [inline scripts/hostapd/uninstall]: Does this need `|| true` when `-rf` is used?

File: scripts/ipfs/install [hunk 8]
```diff
@@ -18,29 +30,43 @@ sudo chown root:staff /usr/local/bin/ipfs
 rm -rf "$BASE_DIR/tmp"
 
 # Initialize IPFS
-ipfs init
+ipfs init || true
```
Comment (benhylau) [inline scripts/ipfs/install]: Does this occasionally fail? Entropy + timeout?

File: scripts/ipfs/ipfs.service [hunk 9]
```diff
@@ -7,7 +7,7 @@ After=network.target
 Type=simple
 Environment=IPFS_PATH=__USER_HOME__/.ipfs
 ExecStart=/usr/local/bin/ipfs daemon --enable-namesys-pubsub --migrate=true
-ExecStartPost=/usr/local/bin/ipfs-swarm-cjdns.sh
+ExecStartPost=/usr/local/bin/ipfs-swarm.sh
```
Comment (benhylau) [inline scripts/ipfs/ipfs.service]: Does this race with yggdrasil and cjdns services?

File: scripts/mesh-adhoc/mesh-adhoc [hunk 10]
```diff
@@ -5,6 +5,12 @@ set -e
 # Set wireless regulatory domain
 sudo iw reg set CA
 
+# Is it safe to stop wpa_supplicant. Is AP running
+if [ ! -d /sys/class/net/wlan-ap ]; then
+  # Kill wpa_supplicant with sometimes conflicts with IBSS Ad-Hoc
```
Comment (benhylau) [inline scripts/mesh-adhoc/mesh-adhoc]: AdHoc / ad-hoc / Ad-Hoc are all used throughout the docs. Pick one :)

File: scripts/mesh-point/mesh-point [hunk 11]
```diff
@@ -6,30 +6,36 @@ function isMeshable {
...
     if  [ -z "$(iw phy phy$1 info | grep 'mesh point')" ]; then
         result='' # Indicate interface is not meshable
-    # XRADIO driver reports Mesh Point support incorrectly (there is no support for it)
-    elif [ "$(basename $(readlink /sys/class/net/$interface/device/driver))" == 'xradio_wlan' ]; then	    
+    # XRADIO driver reports Mesh Point but does not actually work
+    elif [ "$(basename $(readlink /sys/class/net/$interface/device/driver))" == 'xradio_wlan' ]; then
         result='' # Indicate interface is actually not meshable
     else
-        result='1' # Indicate interface is  meshable
+        result='1' # Indicate interface is meshable
```
Comment (benhylau) [inline scripts/mesh-point/mesh-point]:
```suggestion
        result='1' # Indicate interface supports mesh point
```

File: scripts/prometheus-node-exporter/nodestats-tomesh.py [hunk 12]
```diff
@@ -14,76 +15,126 @@
...
-        for int in ints:
-            if int.find("Interface") > -1:
-                    words = int.split()
-                    currentitn=words[1]
-                
-                    # Read mac address from system file
-                    with open("/sys/class/net/" + currentitn + "/address") as f:
-                        mac = f.readlines()
-                    mac = [x.strip() for x in mac][0]
+                #Wireless Link Dump
```
Comment (benhylau) [inline scripts/prometheus-node-exporter/nodestats-tomesh.py]:
```suggestion
                # Wireless Link Dump
```

File: scripts/ssb-web-pi/install [hunk 13]
```diff
@@ -22,27 +23,27 @@ ssbPath="$HOME/.ssb"
 
 # Install broadcast service for users of sbot
 currentUser=$USER
-sudo cp "$BASE_DIR/ssb-web-broadcast-service.sh" "/usr/local/bin/ssb-web-broadcast-service.sh"
-sudo cp "$BASE_DIR/ssb-web-broadcast.service" /etc/systemd/system/ssb-web-broadcast.service
-sudo sed -i "s|__USER__|${currentUser}|g" /etc/systemd/system/ssb-web-broadcast.service
+sudo cp "$BASE_DIR/ssb-web-pi-broadcast-service.sh" "/usr/local/bin/ssb-web-pi-broadcast-service.sh"
```
Comment (benhylau) [inline scripts/ssb-web-pi/install]: Is this needed for patchfoo as well?

File: scripts/ssb-web-pi/ssb-web-pi.conf [hunk 14]
```diff
@@ -1,12 +1,12 @@
-location /sbot {
+location /ssb-web-pi {
     alias /var/www/sbot;
```
Comment (benhylau) [inline scripts/ssb-web-pi/ssb-web-pi.conf]: Is this alias still valid?

File: scripts/yggdrasil-iptunnel/yggdrasil-setup [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    sudo route add -net "${SUBNET4}0/24" ygg0 || true
+
+    # If no NAT (masquarading) being done, start doing it
+    if [ -z "$(sudo iptables -L POSTROUTING -v -n -t nat  | grep MASQUERADE)" ]; then
+        sudo iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
+    fi
+# Configure as yggdrasil iptunnel client if client file is present (this is simply a newline-separated list
+# of cjdns public keys in /etc/yggdrasil.iptunnel.client, each key indicating an iptunnel exit server)
+elif [ -e /etc/yggdrasil.iptunnel.client ]; then
+
```
Comment (benhylau) [inline scripts/yggdrasil-iptunnel/yggdrasil-setup]:
```suggestion
```

File: docs/FAQ.md (hunk 6 above)
Comment (benhylau) [reply docs/FAQ.md]: Not blocking release, can leave as is.

File: scripts/hostapd/uninstall (hunk 7 above)
Comment (darkdrgn2k) [reply scripts/hostapd/uninstall]: In case files does not exist for some reason.

File: scripts/ssb-web-pi/ssb-web-pi.conf (hunk 14 above)
Comment (darkdrgn2k) [reply scripts/ssb-web-pi/ssb-web-pi.conf]: I didn't want to change the actual code due to lost of hard coded things in the code stack. Module is very ALPHA still but has some interesting concepts.

Didnt want to break things.

File: scripts/ipfs/install (hunk 8 above)
Comment (makeworld-the-better-one) [reply scripts/ipfs/install]: I believe it's in case the `~/.ipfs` folder already exists. On my laptop:
```
❯ ipfs init
initializing IPFS node at /home/makeworld/.ipfs
Error: ipfs configuration file already exists!
Reinitializing would overwrite your keys.
```

File: scripts/hostapd/uninstall (hunk 7 above)
Comment (benhylau) [reply scripts/hostapd/uninstall]: I mean this:

```
b@deprecated:~$ rm -r stuff
rm: cannot remove 'stuff': No such file or directory
b@deprecated:~$ rm -rf stuff
b@deprecated:~$ echo $?
0
```

It always returns success with the `-f` :)

File: scripts/mesh-adhoc/mesh-adhoc (hunk 10 above)
Comment (benhylau) [reply scripts/mesh-adhoc/mesh-adhoc]: Filed https://github.com/tomeshnet/prototype-cjdns-pi/issues/318

File: docs/FAQ.md (hunk 6 above)
Comment (benhylau) [reply docs/FAQ.md]: Filed https://github.com/tomeshnet/prototype-cjdns-pi/issues/318

File: scripts/hostapd/uninstall (hunk 7 above)
Comment (benhylau) [reply scripts/hostapd/uninstall]: Not blocking, but maybe remove it in the future since it's unnecessary?

Comment (darkdrgn2k) [reply scripts/hostapd/uninstall]: hmmm interesting
